### PR TITLE
DRILL-3637: One more pull request for ElasticSearch storage plugin

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-3637](https://issues.apache.org/jira/browse/DRILL-3637): ElasticSearch storage plugin implementation based on Calcite adapter

## Description

This PR introduces base classes for creating storage plugins based on Calcite adapters and also provides a plugin based on  Calcite's ElasticSearch adapter. 
This implementation supports
- project pushdown;
- filters pushdown (only predicates supported by Calcite adapter);
- limit pushdown;
- sort pushdown;
- aggregation pushdown;
- schema provisioning.

## Documentation
Docs on Drill web site should be updated to state about the support of Elasticsearch plugin.

## Testing
Added unit tests, checked with ES 6.3.2, 7.8, 7.10.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/DrillDataContext.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.apache.calcite.DataContext;
+import org.apache.calcite.adapter.java.JavaTypeFactory;
+import org.apache.calcite.linq4j.QueryProvider;
+import org.apache.calcite.schema.SchemaPlus;
+
+import java.util.Map;
+
+public class DrillDataContext implements DataContext {
+  private final SchemaPlus rootSchema;
```
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/DrillDataContext.java:28]: Why not use `DataContextImpl` with `CalciteSchema` schema wrapper?
Is it due to `DataContextImpl` limited to JDBC only in the Calicte?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/plan/VertexDrel.java [hunk 2]
```diff
@@ -27,19 +25,21 @@
 import org.apache.drill.exec.planner.logical.DrillImplementor;
 import org.apache.drill.exec.planner.logical.DrillRel;
 
-public class JdbcDrel extends SingleRel implements DrillRel {
+import java.util.List;
+
+public class VertexDrel extends SingleRel implements DrillRel {
```
Comment (vdiravka) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/plan/VertexDrel.java:30]: Could you describe what Vertex Drel means?

File: exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/convert/TestDirectConverter.java [hunk 3]
```diff
@@ -708,7 +708,7 @@ public void testBasicConversionType() {
     expect(ConversionType.IMPLICIT, conversions.analyze(tinyIntCol, bigIntCol));
     expect(ConversionType.IMPLICIT, conversions.analyze(tinyIntCol, float4Col));
     expect(ConversionType.IMPLICIT, conversions.analyze(tinyIntCol, float8Col));
-    expect(ConversionType.IMPLICIT, conversions.analyze(tinyIntCol, decimalCol));
+    expect(ConversionType.EXPLICIT, conversions.analyze(tinyIntCol, decimalCol));
```
Comment (vdiravka) [inline exec/java-exec/src/test/java/org/apache/drill/exec/physical/impl/scan/convert/TestDirectConverter.java:711]: Why it is changed?

**Review by vdiravka — COMMENTED** [review]
Great job!

My review is mostly the questions and small tips. I am fine for very short answers :) 
Thanks for PR!

**Review by vvysotskyi — COMMENTED** [review]
@vdiravka, thanks for the review, I've answered your questions, and added some JavaDocs. Please take a look.

**Review by cgivre — APPROVED** [review]
@vvysotskyi 
Thanks for your work on this!  Per our conversation, I'm good with this with the updated Javadocs.  Can you please add a `README` to the plugin then +1 from me.

**Comment by vvysotskyi** [comment]
@cgivre, I've added `README.md`, please review.

**Comment by cgivre** [comment]
LGTM +1!

**Comment by cgivre** [comment]
@dzamo 
We have a big PR for you to add to the docs!

File: contrib/storage-elasticsearch/pom.xml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <configuration>
+          <forkCount combine.self="override">1</forkCount>
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply contrib/storage-elasticsearch/pom.xml:87]: ok. We can consider ES tests parallelization it in future Drill improvements

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java [hunk 5]
```diff
@@ -284,7 +284,7 @@ public static boolean isLimit0(RexNode fetch) {
   public static boolean isProjectOutputRowcountUnknown(Project project) {
     for (RexNode rex : project.getProjects()) {
       if (rex instanceof RexCall) {
-        if ("flatten".equals(((RexCall) rex).getOperator().getName().toLowerCase())) {
+        if ("flatten".equalsIgnoreCase(((RexCall) rex).getOperator().getName())) {
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/DrillRelOptUtil.java:287]: It's fine. I thought `toLowerCase` is faster than `equalsIgnoreCase`, but no - the last is faster in most cases:
https://stackoverflow.com/questions/29785600/efficiency-of-equalsignorecase-versus-touppercase-equals-and-tolowercase-e

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/AnalyzeTableHandler.java [hunk 6]
```diff
@@ -210,16 +209,6 @@ protected DrillRel convertToDrel(RelNode relNode, AbstractSchema schema, String
...
-    for (RelNode rel : rels) {
-      if (rel instanceof DrillScanRel) {
-        return (DrillScanRel) rel;
-      } else {
-        return findScan(rel.getInputs().toArray(new RelNode[0]));
-      }
-    }
-    return null;
-  }
   // Make sure no unsupported features in ANALYZE statement are used
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/AnalyzeTableHandler.java:212]: Sure

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/SubsetRemover.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store;
+
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelShuttle;
+import org.apache.calcite.rel.RelShuttleImpl;
+
+public class SubsetRemover extends RelShuttleImpl {
```
[the remark this answers is not in the document]
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/SubsetRemover.java:28]: Could you add please info why and when we may want removing `RelSubset`. Thanks

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/DrillDataContext.java (hunk 1 above)
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/DrillDataContext.java:28]: Possibly can be considered for Calcite improvement?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/plan/VertexDrel.java (hunk 2 above)
Comment (vdiravka) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/enumerable/plan/VertexDrel.java:34]: Nice

**Review by vdiravka — APPROVED** [review]
+1 LGTM

**Pull request merged**